### ISSUE TITLE
Signed LD calculations

### ISSUE THREAD

**Issue by vsbuffalo** [tskit-dev/tskit#2077]
The new neat `LdCalculator` class has methods for r2 matrices and arrays. Could I request that signed LD (i.e. the covariance between gametic states) be added as a method? This is rather important in some selection scenarios.

**Comment by jeromekelleher** [comment]
The LDCalculator class is actually very old @vsbuffalo, and scheduled for deprecation/removal (#1900). The idea is that we replace it with some TreeSequence methods that will compute the same values.

The medium term plan is to implement the general framework that @apragsdale devised (prototyped in #432) so that we can have a wide variety of multi-site stats, following the same general patterns as the single-site stat framework.

In the short term though, we could definitely widen the set of LD-stats that we're computing easily enough, which might be a helpful thing to do in preparation for the general framework. Adding "r" and "D" like stats to the current setup should be easy enough I think, but we should probably generalise to multiple mutations before doing that.

If you'd like to help out here, I'd be happy to discuss some details.

**Comment by vsbuffalo** [comment]
Thanks for the clarification, Jerome! I hard heard there was new LD code coming and I think I got my wires crossed on what was what. I look forward to these new additions! A temporary D estimation routine would be useful (though I suppose I could hack my own together).

**Comment by jeromekelleher** [comment]
There's not really anyone to work on this unfortunately @vsbuffalo, we're behind on a lot of projects and there's not sufficient motivation. As I say, I'm happy to give some pointers if you're interested in helping out, though.

**Comment by vsbuffalo** [comment]
For now, running `np.cov()` on a haplotype matrix is all I need! Sharing this in case it's helpful to others.

**Comment by twooldridge** [comment]
Hi Vince,

I'm interested in LD calculations as well, I'm wondering if you can clarify the tree sequence -> haplotype matrix -> np.cov() sequence of commands you used? Thanks!

Best,
Brock

**Comment by apragsdale** [comment]
Hi Brock - Since this issue was opened, the LD calculation methods have expanded. You can now compute some additional statistics, like signed LD.

For example:
```python
r2_matrix = ts.ld_matrix(stat="r2") # squared correlation
r_matrix = ts.ld_matrix(stat="r") # correlation (signed)
D_matrix = ts.ld_matrix(stat="D") # covariance (signed)
```

@lkirk has been developing these LD methods, and there are some additional features as well. Hope this helps!

**Comment by hyanwong** [comment]
Is there an issue open to document `ts.ld_matrix`? Is it stable enough to want to expose it now?

**Comment by apragsdale** [comment]
That's a good question. It is released at this point, in tskit-0.6, but you're right that it isn't documented. There is work to do still for LD scores and LD decay calculations, but I believe the LD matrix API is just about there. I'll let @lkirk chime in and correct me...

Should I hold off from pointing people to this code until it's fully documented?

**Comment by jeromekelleher** [comment]
I haven't really kept up to date to be honest - if the LD calculations are ready for production then we should document them I guess. It would be good to characterise the performance profiles though, as the LdCalculator is optimised for a different regime, as I understand it.
